pkg/microservice/aslan/core/environment/service/environment.go

```diff
@@ -601,7 +596,28 @@ func UpdateProduct(existedProd, updateProd *commonmodels.Product, renderSet *com
 			err = e.ErrUpdateEnv.AddDesc(err.Error())
 			return
 		}
-		err = commonrepo.NewProductColl().UpdateGroup(envName, productName, groupIndex, groupServices)
+		var updateGroup []*commonmodels.ProductService
+		newServiceMap := make(map[string]*commonmodels.ProductService)
+		for _, service := range groupServices {
+			newServiceMap[service.ServiceName] = service
+		}
+		oldServiceMap := make(map[string]*commonmodels.ProductService)
+		for _, serviceGroup := range existedProd.Services {
+			for _, service := range serviceGroup {
+				oldServiceMap[service.ServiceName] = service
+				if newService, ok := newServiceMap[service.ServiceName]; ok {
+					updateGroup = append(updateGroup, newService)
+					continue
+				}
+				updateGroup = append(updateGroup, service)
+			}
+		}
+		for _, newService := range groupServices {
+			if _, ok := oldServiceMap[newService.ServiceName]; !ok {
+				updateGroup = append(updateGroup, newService)
+			}
+		}
```

Comment:
the logic of generate udateGroup can be optimized:
1. use `prodServiceGroup` instead of use a nested loop of existedProd.Services
2. build a map[string]*ProductService, fill them with services in existedProd and then fill with services in  groupServices again(if i understand correctly)
3. add some comments, this function is really complicated